packages/auth-nextjs/src/shared.ts

```diff
@@ -138,6 +137,107 @@ export abstract class NextAuth extends NextAuthHelpers {
     });
   }
 
+  public oAuth = {
+    /**
+     * Start the OAuth flow by getting the OAuth configuration from the request
+     * URL search parameters and redirecting to the auth extension server's
+     * authorize endpoint.
+     *
+     * @param {NextRequest} req The incoming request.
+     * @param {string=} req.nextUrl.searchParams.provider_name The name of the
+     * OAuth provider to use.
+     * @param {string=} req.nextUrl.searchParams.authorize_url The URL to
+     * redirect to to start the OAuth flow. If not provided, will default to
+     * the auth extension server's authorize endpoint.
+     * @param {string=} req.nextUrl.searchParams.callback_url The URL to
+     * redirect to within the OAuth flow once the user has authorized the OAuth
+     * client.
+     */
+    handleOAuth: async (req: NextRequest): Promise<NextResponse> => {
+      const providerName = req.nextUrl.searchParams.get("provider_name");
+      if (!providerName) {
+        throw new InvalidDataError("Missing provider_name in request");
+      }
+
+      const callbackUrl = req.nextUrl.searchParams.get("callback_url");
+
+      const authBasePath = this._authRoute.endsWith("/")
+        ? this._authRoute
+        : `${this._authRoute}/`;
+      const redirectTo = new URL("oauth/callback", authBasePath);
+      const redirectToOnSignup = new URL(redirectTo);
+      redirectToOnSignup.searchParams.set("isSignUp", "true");
+
+      const authorizeUrl =
+        req.nextUrl.searchParams.get("authorize_url") ??
```

Comment:
Should this (and `callback_url`) be args to this `handleOAuth` function? I'm not sure they are meant to be something the user should be able to control? and in the example usage thay're appended to the req right before it's passed to this function anyway.

Reply:
The API I'm going for here is that these "handler" functions should always take a `NextRequest` and return a `Promise<NextResponse>` in a middleware like fashion. An alternative way to do this is to be more explicit about what parameters we explicitly depend on within the function itself, but I'm worried that it adds boilerplate of getting the data we need from the request. I can sketch that out though and see how that feels in actual use though.

Reply:
Ok, maybe strip out those two values from the search params here: https://github.com/edgedb/edgedb-js/pull/1136/files#diff-660934b707fc817f8bd5cef3341ffb7f24ffc16f9e200843c762e384e2820c41L173 then? I haven't fully thought out the whole flow and there probably isn't an attack you could do using these params, but just be safe maybe 😄.

Reply:
> Ok, maybe strip out those two values from the search params here: [#1136 (files)](https://github.com/edgedb/edgedb-js/pull/1136/files#diff-660934b707fc817f8bd5cef3341ffb7f24ffc16f9e200843c762e384e2820c41L173) then?

Maybe I'm not understanding the suggestion but these parameters _will_ be present in the URL that the browser uses as a link, just like `provider`, and it needs to be passed all the way to the auth extension. It's not something an _end user_ needs to worry about, but the application developer will absolutely need to set this either in the link button's URL, or manually in their own backend like I did in my example.

> I haven't fully thought out the whole flow and there probably isn't an attack you could do using these params, but just to be safe

The URLs are checked against the `allowed_redirect_urls` config, so if there is an attack vector there, the attack also exists in the existing `redirect_to` search parameters.